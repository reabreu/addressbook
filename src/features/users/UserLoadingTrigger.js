import React from "react";
import { Loading, Loader } from "./styles";

export default ({ element, showLoader }) =>
  showLoader ? (
    <Loading ref={element}>
      <Loader />
    </Loading>
  ) : null;
