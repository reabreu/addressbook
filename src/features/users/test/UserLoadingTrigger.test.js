import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import getStore from "../../../store";
import UserLoadingTrigger from "../UserLoadingTrigger";
import { fullStoreEN } from "./mock/full-store-en";

function setupIntersectionObserverMock({
  observe = () => null,
  unobserve = () => null,
} = {}) {
  class IntersectionObserver {
    observe = observe;
    unobserve = unobserve;
    disconnect = unobserve;
  }

  Object.defineProperty(window, "IntersectionObserver", {
    writable: true,
    configurable: true,
    value: IntersectionObserver,
  });
  Object.defineProperty(global, "IntersectionObserver", {
    writable: true,
    configurable: true,
    value: IntersectionObserver,
  });
}

describe("<UserList /> spec", () => {
  beforeEach(() => {
    setupIntersectionObserverMock();
  });

  afterEach(() => {
    window.IntersectionObserver = undefined;
  });

  it("Matches initial snapshot with no users and loading spinner", () => {
    const store = getStore();
    const { asFragment } = render(
      <Provider store={store}>
        <UserLoadingTrigger />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("Shows message when current lang is full", () => {
    const store = getStore(fullStoreEN);
    const { asFragment } = render(
      <Provider store={store}>
        <UserLoadingTrigger />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("Shows button on error to allow manual fetch", () => {
    const store = getStore({
      ...fullStoreEN,
      users: {
        values: [],
        status: "error",
      },
    });
    const { asFragment } = render(
      <Provider store={store}>
        <UserLoadingTrigger />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
