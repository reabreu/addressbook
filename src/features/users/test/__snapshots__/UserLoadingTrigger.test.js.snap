// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserList /> spec Matches initial snapshot with no users and loading spinner 1`] = `
<DocumentFragment>
  <div
    class="sc-AxgMl jOotEa"
  >
    <div
      class="sc-Axmtr bFiunQ"
    />
  </div>
</DocumentFragment>
`;

exports[`<UserList /> spec Shows button on error to allow manual fetch 1`] = `
<DocumentFragment>
  <div
    class="sc-AxgMl jOotEa"
  >
    <button>
      There was an error fetching users, click here to try again
    </button>
  </div>
</DocumentFragment>
`;

exports[`<UserList /> spec Shows message when current lang is full 1`] = `
<DocumentFragment>
  <div
    class="sc-AxgMl jOotEa"
  >
    <label>
      End of users catalog
    </label>
  </div>
</DocumentFragment>
`;
